scapy/contrib/isotp/isotp_soft_socket.py

```diff
@@ -280,104 +265,103 @@ def cancel(handle):
 
                 if len(handles) == 0:
                     # set the event to stop the wait - this kills the thread
-                    TimeoutScheduler._event.set()
+                    cls._event.set()
             else:
                 raise Scapy_Exception("Handle not found")
 
-    @staticmethod
-    def clear():
+    @classmethod
+    def clear(cls):
         # type: () -> None
         """Cancels the execution of all timeouts."""
-        with TimeoutScheduler._mutex:
-            TimeoutScheduler._handles.clear()
+        with cls._mutex:
+            cls._handles = []
 
         # set the event to stop the wait - this kills the thread
-        TimeoutScheduler._event.set()
+        cls._event.set()
 
-    @staticmethod
-    def _peek_next():
+    @classmethod
+    def _peek_next(cls):
         # type: () -> Optional[TimeoutScheduler.Handle]
         """Returns the next timeout to execute, or `None` if list is empty,
         without modifying the list"""
-        with TimeoutScheduler._mutex:
-            handles = TimeoutScheduler._handles
+        with cls._mutex:
+            handles = cls._handles
             if len(handles) == 0:
                 return None
             else:
                 return handles[0]
 
-    @staticmethod
-    def _wait(handle):
+    @classmethod
+    def _wait(cls, handle):
         # type: (Optional[TimeoutScheduler.Handle]) -> None
         """Waits until it is time to execute the provided handle, or until
         another thread calls _event.set()"""
 
         if handle is None:
-            when = TimeoutScheduler.GRACE
+            when = cls.GRACE
         else:
             when = handle._when
 
         # Check how much time until the next timeout
-        now = TimeoutScheduler._time()
-        to_wait = when - now
+        now = cls._time()
+        to_wait = min(when - now, cls.GRACE)
 
         # Wait until the next timeout,
         # or until event.set() gets called in another thread.
         if to_wait > 0:
             log_runtime.debug("TimeoutScheduler Thread going to sleep @ %f " +
                               "for %fs", now, to_wait)
-            interrupted = TimeoutScheduler._event.wait(to_wait)
-            new = TimeoutScheduler._time()
+            interrupted = cls._event.wait(to_wait)
```

Comment:
@gpotter2 This line made all the instability. Sometimes, in raw cases, this event didn't got interrupted.